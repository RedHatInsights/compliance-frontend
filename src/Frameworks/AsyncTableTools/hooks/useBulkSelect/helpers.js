import React from 'react';
import { Spinner } from '@patternfly/react-core';

export const compileTitle = (itemsTotal, loading) => {
  if (loading === true) {
    return <Spinner size="sm" />;
  } else return `${itemsTotal} selected`;
};

export const checkboxState = (selectedItemsTotal, itemsTotal) => {
  if (selectedItemsTotal === 0) {
    return false;
  } else if (selectedItemsTotal === itemsTotal) {
    return true;
  } else {
    return null;
  }
};

export const selectOrUnselect = (selected) =>
  selected ? 'Unselect' : 'Select';

const allItemsIncluded = (items = [], selection = []) =>
  items.filter((item) => selection.includes(item)).length === items.length;

export const checkCurrentPageSelected = (items = [], selectedItems = []) => {
  if (selectedItems.length === 0) {
    return false;
  } else {
    return allItemsIncluded(items, selectedItems);
  }
};

export const itemIds = (items) => items.map(({ itemId }) => itemId);

export const selectItemTransformer = (item, selectedIds = []) => ({
  ...item,
  rowProps: {
    selected: selectedIds.includes(item.itemId),
  },
});
