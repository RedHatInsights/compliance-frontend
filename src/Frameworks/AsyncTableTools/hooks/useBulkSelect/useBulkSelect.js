import { useCallback } from 'react';
import useSelectionManager from '../useSelectionManager';
import useFetchBatched from '@redhat-cloud-services/frontend-components-utilities/useFetchBatched';
import {
  checkCurrentPageSelected,
  checkboxState,
  compileTitle,
  selectOrUnselect,
} from './helpers';

/**
 *  @typedef {object} useBulkSelectReturn
 *
 *  @property {Function} [markRowSelected] "Transformer" function to be passed to the rowsBuilder
 *  @property {object}   [toolbarProps]    Object containing PrimaryToolbar props
 *  @property {object}   [tableProps]      Object containing Patternfly (v4) Table props
 *
 */

/**
 * Provides properties for a Pattternfly (based) Table and Toolbar component to implement bulk selection
 *
 *  @param   {object}              [options]                AsyncTableTools options
 *  @param   {number}              [options.total]          Number to show as total count
 *  @param   {Function}            [options.onSelect]       function to call when a selection is made
 *  @param   {Array}               [options.preselected]    Array of itemIds selected when initialising
 *  @param   {Function}            [options.itemIdsInTable] Function to call to retrieve IDs when "Select All" is chosen
 *  @param   {Array}               [options.itemIdsOnPage]  Array of item ids visible on the page
 *  @param   {string}              [options.identifier]     Property of the items that should be used as ID to select them
 *
 *  @returns {useBulkSelectReturn}                          Functions and props to use for setting up bulk selection
 *
 *  @category AsyncTableTools
 *  @subcategory Hooks
 *
 */
const useBulkSelect = ({
  total = 0,
  onSelect,
  preselected,
  itemIdsInTable,
  itemIdsOnPage,
  identifier = 'itemId',
}) => {
  const enableBulkSelect = !!onSelect;
  const {
    selection: selectedIds,
    set,
    select,
    deselect,
    clear,
  } = useSelectionManager(preselected);
  const selectedIdsTotal = (selectedIds || []).length;
  const paginatedTotal = itemIdsOnPage?.length || total;
  const allSelected = selectedIdsTotal === total;
  const noneSelected = selectedIdsTotal === 0;
  const currentPageSelected = checkCurrentPageSelected(
    itemIdsOnPage,
    selectedIds
  );

  const isDisabled = total === 0;
  const checked = checkboxState(selectedIdsTotal, total);

  const { isLoading, fetchBatched } = useFetchBatched();
  const title = compileTitle(selectedIdsTotal, isLoading);

  const selectOne = useCallback(
    (_, _selected, _key, row) =>
      row.selected ? deselect(row[identifier]) : select(row[identifier]),
    [select, deselect, identifier]
  );
  const selectPage = useCallback(
    () =>
      !currentPageSelected ? select(itemIdsOnPage) : deselect(itemIdsOnPage),
    [select, deselect, itemIdsOnPage, currentPageSelected]
  );

  const selectAll = async () => {
    if (allSelected) {
      clear();
    } else {
      set(await fetchBatched(itemIdsInTable, total));
    }
  };

  const markRowSelected = (row) => ({
    ...row,
    isSelected: selectedIds.includes(row.itemId),
  });

  return enableBulkSelect
    ? {
        markRowSelected,
        tableProps: {
          onSelect: total > 0 ? selectOne : undefined,
          canSelectAll: false,
        },
        toolbarProps: {
          bulkSelect: {
            toggleProps: { children: [title] },
            isDisabled,
            items: [
              {
                title: 'Select none',
                onClick: clear,
                props: {
                  isDisabled: noneSelected,
                },
              },
              ...(itemIdsOnPage
                ? [
                    {
                      title: `${selectOrUnselect(
                        currentPageSelected
                      )} page (${paginatedTotal} items)`,
                      onClick: selectPage,
                    },
                  ]
                : []),
              ...(itemIdsInTable
                ? [
                    {
                      title: `${selectOrUnselect(
                        allSelected
                      )} all (${total} items)`,
                      onClick: selectAll,
                    },
                  ]
                : []),
            ],
            checked,
            onSelect: !isDisabled ? selectPage : undefined,
          },
        },
      }
    : {};
};

export default useBulkSelect;
