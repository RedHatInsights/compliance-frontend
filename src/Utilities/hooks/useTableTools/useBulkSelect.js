import { useEffect, useState } from 'react';
import { filterSelected } from './helper';

const compileTitle = (itemsTotal, titleOption) => {
  if (typeof titleOption === 'string') {
    return titleOption;
  } else if (typeof titleOption === 'function') {
    return titleOption(itemsTotal);
  } else {
    return `${itemsTotal} selected`;
  }
};

const checkboxState = (selectedItemsTotal, itemsTotal) => {
  if (selectedItemsTotal === 0) {
    return false;
  } else if (selectedItemsTotal === itemsTotal) {
    return true;
  } else {
    return null;
  }
};

const allItemsIncluded = (items, selection = []) =>
  items.filter((item) => selection.includes(item)).length === items.length;

const selectOrUnselect = (selected) => (selected ? 'Unselect' : 'Select');

const checkCurrentPageSelected = (items, selectedItems) => {
  if (selectedItems.length === 0) {
    return false;
  } else {
    return allItemsIncluded(items, selectedItems);
  }
};

const itemIds = (items) => items.map((item) => item.itemId);
const mergeArraysUniqly = (arrayA, arrayB) =>
  Array.from(new Set([...arrayA, ...arrayB]));

export const useBulkSelect = ({
  total,
  onSelect,
  preselected,
  itemIdsInTable,
  itemIdsOnPage,
  identifier = 'id',
  showTreeTable,
}) => {
  const enableBulkSelect = !!onSelect;
  const [selectedIds, setSelectedItemIds] = useState(preselected);
  const selectedIdsTotal = (selectedIds || []).length;
  const paginatedTotal = itemIdsOnPage().length;
  const allSelected = selectedIdsTotal === total;
  const noneSelected = selectedIdsTotal === 0;
  const currentPageSelected = checkCurrentPageSelected(
    itemIdsOnPage(),
    selectedIds || []
  );

  const isDisabled = total === 0;
  const checked = checkboxState(selectedIdsTotal, total);
  const title = compileTitle(selectedIdsTotal);

  const onSelectCallback = async (func) => {
    const newSelectedItemsIds = await func();
    setSelectedItemIds(newSelectedItemsIds);
    onSelect && onSelect(newSelectedItemsIds);
  };

  const selectItems = (itemIds) => mergeArraysUniqly(selectedIds, itemIds);

  const unselectItems = (itemIds) =>
    selectedIds.filter((itemId) => !itemIds.includes(itemId));

  const unselectAll = () => [];
  const selectNone = () => onSelectCallback(unselectAll);
  const selectOne = (_, selected, key, row) => {
    onSelectCallback(() =>
      selected
        ? selectItems([row[identifier]])
        : unselectItems([row[identifier]])
    );
  };

  const selectPage = () =>
    onSelectCallback(() => {
      const currentPageIds = itemIdsOnPage();
      const currentPageSelected =
        mergeArraysUniqly(selectedIds, currentPageIds).length ===
        selectedIds.length;

      return currentPageSelected
        ? unselectItems(currentPageIds)
        : selectItems(currentPageIds);
    });

  const selectAll = () =>
    onSelectCallback(async () =>
      allSelected ? unselectAll() : selectItems(await itemIdsInTable())
    );

  useEffect(() => {
    setSelectedItemIds(preselected);
  }, [preselected]);

  return enableBulkSelect
    ? {
        selectedIds,
        selectNone,
        selectItems: (ids) => onSelectCallback(() => selectItems(ids)),
        unselectItems: (ids) => onSelectCallback(() => unselectItems(ids)),
        tableProps: {
          ...(!showTreeTable
            ? { onSelect: total > 0 ? selectOne : undefined }
            : {}),
          canSelectAll: false,
        },
        toolbarProps: {
          bulkSelect: {
            toggleProps: { children: [title], count: total },
            isDisabled,
            items: [
              {
                title: 'Select none',
                onClick: selectNone,
                props: {
                  isDisabled: noneSelected,
                },
              },
              {
                title: `${selectOrUnselect(
                  currentPageSelected
                )} page (${paginatedTotal} items)`,
                onClick: selectPage,
              },
              {
                title: `${selectOrUnselect(allSelected)} all (${total} items)`,
                onClick: selectAll,
              },
            ],
            checked,
            onSelect: !isDisabled ? selectPage : undefined,
          },
        },
      }
    : {};
};

const selectItemTransformer = (item, selectedIds) => ({
  ...item,
  rowProps: {
    selected: (selectedIds || []).includes(item.itemId),
  },
});

export const useBulkSelectWithItems = ({
  onSelect,
  items: propItems,
  filter,
  paginator,
  preselected,
  setPage,
  ...options
}) => {
  const enableBulkSelect = !!onSelect;
  const items = propItems.map((item) =>
    selectItemTransformer(item, preselected)
  );
  const total = items.length;

  const filteredItems = filter ? filter(items) : items;
  const filteredTotal = filteredItems.length;
  const filtered = filteredTotal !== total;

  const paginatedItems = paginator ? paginator(filteredItems) : filteredItems;
  const paginatedTotal = paginatedItems.length;

  const allCount = filtered ? filteredTotal : total;

  useEffect(() => {
    if (paginatedTotal === 0 && setPage) {
      setPage?.(-1);
    }
  }, [paginatedTotal]);

  const { selectNone, selectedIds, ...bulkSelect } = useBulkSelect({
    ...options,
    total: allCount,
    onSelect,
    preselected,
    itemIdsInTable: () => (filtered ? itemIds(filteredItems) : itemIds(items)),
    itemIdsOnPage: () => itemIds(paginatedItems),
    identifier: 'itemId',
  });

  return enableBulkSelect
    ? {
        transformer: (item) => selectItemTransformer(item, selectedIds),
        selectedItems: filterSelected(items, selectedIds),
        selected: selectedIds,
        clearSelection: selectNone,
        ...bulkSelect,
      }
    : {};
};
