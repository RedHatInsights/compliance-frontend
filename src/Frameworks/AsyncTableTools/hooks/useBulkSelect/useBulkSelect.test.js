import useBulkSelect from './useBulkSelect';
import { renderHook } from '@testing-library/react';

describe('useBulkSelect', () => {
  const defaultOptions = {
    total: 0,
    onSelect: () => ({}),
    itemIdsInTable: () => [],
    itemIdsOnPage: [],
  };

  it('returns a bulk select configuration', () => {
    const { result } = renderHook(() => useBulkSelect(defaultOptions));

    expect(result).toMatchSnapshot();
  });

  it('returns a bulk select configuration without select all', () => {
    const { result } = renderHook(() =>
      useBulkSelect({
        ...defaultOptions,
        total: 2,
        preselected: ['ID'],
        itemIdsInTable: () => {
          return ['ID', 'ID1'];
        },
        itemIdsOnPage: ['ID', 'ID1'],
      })
    );

    expect(result.current.toolbarProps.bulkSelect.items).toMatchSnapshot();
  });

  it('returns a bulk select configuration with select all', () => {
    const { result } = renderHook(() =>
      useBulkSelect({
        ...defaultOptions,
        total: 2,
        preselected: ['ID'],
        fetchAll: Promise.resolve(['2417de', '51b20a']),
        itemIdsOnPage: ['ID', 'ID1'],
      })
    );

    expect(result.current.toolbarProps.bulkSelect.items).toMatchSnapshot();
  });

  it('returns a bulk select configuration with 1 selected item', () => {
    const { result } = renderHook(() =>
      useBulkSelect({
        ...defaultOptions,
        total: 2,
        preselected: ['ID'],
        itemIdsInTable: () => ['ID', 'ID2'],
        itemIdsOnPage: ['ID'],
      })
    );

    // eslint-disable-next-line testing-library/no-node-access
    expect(result.current.toolbarProps.bulkSelect.toggleProps.children).toEqual(
      ['1 selected']
    );
  });
});
