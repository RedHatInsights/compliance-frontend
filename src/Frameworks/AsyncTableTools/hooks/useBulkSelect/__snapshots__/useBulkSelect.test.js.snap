// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useBulkSelect returns a bulk select configuration 1`] = `
{
  "current": {
    "markRowSelected": [Function],
    "tableProps": {
      "canSelectAll": false,
      "onSelect": undefined,
    },
    "toolbarProps": {
      "bulkSelect": {
        "checked": false,
        "isDisabled": true,
        "items": [
          {
            "onClick": [Function],
            "props": {
              "isDisabled": true,
            },
            "title": "Select none",
          },
          {
            "onClick": [Function],
            "title": "Select page (0 items)",
          },
          {
            "onClick": [Function],
            "title": "Unselect all (0 items)",
          },
        ],
        "onSelect": undefined,
        "toggleProps": {
          "children": [
            "0 selected",
          ],
        },
      },
    },
  },
}
`;

exports[`useBulkSelect returns a bulk select configuration with select all 1`] = `
[
  {
    "onClick": [Function],
    "props": {
      "isDisabled": false,
    },
    "title": "Select none",
  },
  {
    "onClick": [Function],
    "title": "Select page (2 items)",
  },
  {
    "onClick": [Function],
    "title": "Select all (2 items)",
  },
]
`;

exports[`useBulkSelect returns a bulk select configuration without select all 1`] = `
[
  {
    "onClick": [Function],
    "props": {
      "isDisabled": false,
    },
    "title": "Select none",
  },
  {
    "onClick": [Function],
    "title": "Select page (2 items)",
  },
  {
    "onClick": [Function],
    "title": "Select all (2 items)",
  },
]
`;
